storage: Disable TS for `__consumer_offsets`

Lazin opened the pull request:
This PR prevents Tiered Storage from being enabled for `__consumer_offsets` if TS is enabled by default.
It also disables TS for existing topics by altering the output of the `is_archival_enabled` and `is_tiered_storage_enabled` method of the `ntp_config`. This has similar effect to disabling this properties manually for the `__consumer_offsets` topic. The archival STM is no longer created for the CO topic. 

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
### Bug Fixes

* Tiered storage and Iceberg can no longer be enabled for the `__consumer_offsets` topic.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--c877c1af-e666-4c50-93a9-f409f21941e2-->
<!--TEST RESULTS-->
<!--72a77664-bea8-4b0d-a00f-ecfaf9fec1cf-->
<!--TEST RESULTS-->
<!--67d22b7d-61d2-4805-bfdd-3c6d474eaab1-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64321">build#64321</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64321#01961afa-eb91-41f4-ae7e-860fce9ddc71 | FLAKY | 1/4 |
| rptest.tests.maintenance_test.MaintenanceCycleTest.test_leader_distribution.use_rpk=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64321#01961afa-eb92-466d-89a7-7ab4b88d36c3 | FLAKY | 1/2 |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move.replication_factor=3.unclean_abort=True.recovery=restart_recovery.compacted=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64321#01961afa-eb92-466d-89a7-7ab4b88d36c3 | FLAKY | 1/2 |
| rptest.tests.rack_aware_replica_placement_test.RackAwarePlacementTest.test_replica_placement.rack_layout_str=ooooFF.num_partitions=50.replication_factor=3.num_topics=2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64321#01961afa-eb92-4c33-a8f0-2dab5043f5f8 | FLAKY | 1/2 |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group | ducktape | https://buildkite.com/redpanda/redpanda/builds/64321#01961afa-eb92-4c33-a8f0-2dab5043f5f8 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65592">build#65592</a></summary>


| test_id                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65592#0196a7a3-409b-4a74-bc91-f6151deb9e79 | FLAKY         | 12/21    |
| rptest.tests.topic_creation_test.RecreateTopicMetadataTest.test_recreated_topic_metadata_are_valid.replication_factor=3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65592#0196a7be-07c8-4023-a2d4-bdda33300eea | FLAKY         | 19/21    |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65592#0196a7be-07c5-445a-a871-38a5209fb39b | FLAKY         | 10/21    |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68758">build#68758</a></summary>


| test_class                  | test_method                              | test_arguments                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------|:-----------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterConfigTest           | test_valid_settings                      |                                                                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68758#0197f58c-a59a-4e1f-b0f9-8f566cb8c4ae | FLAKY         | 17/21    | upstream reliability is '82.97271872060207'. current run reliability is '80.95238095238095'. drift is 2.02034 and the allowed drift is set to 50. The test should PASS  |
| ConsumerOffsetsRecoveryTest | test_consumer_offsets_partition_recovery |                                                                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68758#0197f589-656a-4fdf-83ef-fccfa6a67a82 | FLAKY         | 20/21    | upstream reliability is '98.75311720698254'. current run reliability is '95.23809523809523'. drift is 3.51502 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest    | test_node_operations                     | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68758#0197f589-656c-4f1f-bff0-44d8250d32f1 | FLAKY         | 20/21    | upstream reliability is '99.51690821256038'. current run reliability is '95.23809523809523'. drift is 4.27881 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest    | test_node_operations                     | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68758#0197f589-656e-4728-b9ef-f75cf9d40a31 | FLAKY         | 19/21    | upstream reliability is '99.12280701754386'. current run reliability is '90.47619047619048'. drift is 8.64662 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest    | test_node_operations                     | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68758#0197f589-6564-4d8b-a5ae-7a59e31e609c | FLAKY         | 20/21    | upstream reliability is '99.72375690607734'. current run reliability is '95.23809523809523'. drift is 4.48566 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest    | test_node_operations                     | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68758#0197f589-6566-4b3a-a876-5d873a441454 | FLAKY         | 20/21    | upstream reliability is '93.00518134715026'. current run reliability is '95.23809523809523'. drift is -2.23291 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

rockwotj commented:
related: https://github.com/redpanda-data/redpanda/issues/15097#issuecomment-1823212708

Lazin commented:
> related: [#15097 (comment)](https://github.com/redpanda-data/redpanda/issues/15097#issuecomment-1823212708)

I didn't see this. @rockwotj do you think that this approach makes sense?

andrwng commented on src/v/storage/ntp_config.h:
```diff
@@ -15,18 +15,64 @@
+    // during the topic creation because cloud topic can't be transformed
+    // into non-cloud topic and vice versa.
+    static ntp_caps
+    make_caps(const model::ntp& ntp, bool cloud_topics_override = false) {
```
It seems odd to pass in an override, given my read on this class is it should be entirely focused on the NTP name. Plus, we check the override inline in `cloud_topic_enabled()` anyway?

Lazin replied:
The idea is that every NTP has a set of inherent capabilities and I want to have them in one entity in the code.
For most topics this is based only on the name of the topic and its namespace. But the cloud topic is different. It's impossible to disable or enable the cloud topic property. You can only set this topic property when the topic is created. After that it can't be altered. So I thought that it'd be better to have this reflected here. Otherwise the NTP could be both TS and CT which is not correct.

andrwng replied:
>You can only set this topic property when the topic is created.

So is ntp_caps supposed to be "based on the name, NTP is eligible for supporting feature X" or "NTP was configured with feature X"? My understanding was that `ntp_caps` encompassed the former. And that if we want something like the latter, it should be a function of the ntp config or topic_properties. Otherwise, it isn't quite clear what conditions belong in ntp_caps and which belong in ntp_configs or elsewhere

andrwng commented on src/v/storage/ntp_config.h:
```diff
@@ -15,18 +15,64 @@
+    bool _cloud_topics;
+};
+
 class ntp_config {
```
Do we need to update `archival_metadata_stm_factory::create()` too?

Lazin replied:
yes

andrwng commented on src/v/storage/ntp_config.h:
```diff
@@ -15,18 +15,64 @@
+                                   && ntp.tp.topic
+                                        == model::kafka_consumer_offsets_topic;
+        // For now only take into account the namespace and topic name.
+        return ntp_caps(
+          !is_consumer_offsets && !is_internal_topic,
+          !is_consumer_offsets && !is_internal_topic,
+          // Once set cloud-topics override can not be changed.
+          !is_consumer_offsets && !is_internal_topic && cloud_topics_override);
```
nit: this is fine, though I'm not particularly sold on precomputing and encapsulating it in a class, vs a set of utilities functions in a single location that are like `name_supportes_tiered_storage(ntp)` and `name_supports_iceberg(ntp)` or somesuch. I found the separation of the make_caps() impl and the ntp_caps class itself a bit awkward to read

Lazin replied:
I want to be able to print this thing and pass it around.

andrwng commented on src/v/storage/ntp_config.h:
```diff
@@ -15,18 +15,64 @@
+    static ntp_caps
+    make_caps(const model::ntp& ntp, bool cloud_topics_override = false) {
+        bool is_internal_topic = ntp.ns == model::kafka_internal_namespace;
+        bool is_consumer_offsets = ntp.ns == model::kafka_namespace
+                                   && ntp.tp.topic
+                                        == model::kafka_consumer_offsets_topic;
```
It would be great to add an test that gets data into tiered storage for the CO topic, and then upgrades to this version, to make sure nothing breaks (e.g. after upgrading, the group manager can still be replayed, no weird anomalies, etc etc)

Lazin replied:
I'll double check but it looks like we have an upgrade test which enables TS and uses CO simultaneously. So it should be covered.

nvartolomei commented:
> https://github.com/redpanda-data/redpanda/pull/25762#discussion_r2038257068

It sounds like we don't need a cloud topic capability at all then? Whether or not something has the cloud topic capability depends entirely on its config and whether a topic was created with that config set. Everything has the cloud topic capability otherwise and whether or not it can be enabled is entirely controlled by the create topic/alter topic config validation.

Tiered Storage and Iceberg are actually quite similar. We want to come up with the right configs during topic creation actually, before we have an `ntp_config` object. I.e. based on namespace/topic and cluster config and optional user overrides we want to compute whether the operation is allowed or not (e.g. for remote read the default being nullopt (i.e. inherit cluster) and for something like consumer offsets it should evaluate as disabled). The ntp_caps as modeled here don't allow for that so I qualify it as a poor abstraction/primitive.

---

For this current requirement (disable TS for CO) we conceptually need to migrate a topic config during the upgrade. The way we try to migrate it here is a hacky override. This hacky override can as well be implemented where we mutate/set the config overrides and force set archival related properties to disabled without introducing any new concepts.

---

Long term view:

- No NTP caps. We validate configs during creation/mutation. Based on NT, based on cluster config, based on enterprise license, based on previous config, based on all new config property values.
- Disabling a feature for consumer offsets is a one-off change and in the absence of proper migration mechanisms/ease of change we implement a hacky override for topic configs.
- - We can probably also introduce a proper migration on controller level and avoid the need for overrides "overrides" at all?

Lazin commented:
@nvartolomei I don't mind getting rid of the `ntp_caps` altogether. But I don't think that doing this at the controller level is feasible.

Lazin committed:
archival: Disable archival STM for consumer offsets

We don't want TS to be enabled for the consumer offsets topic. This
change will make TS treat consumer offsets as internal topic.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
cluster: Do not create ntp_archiver for CO partitions

Treat consumer offsets topic as an internal topic. The archiver will
not be started for it even if the TS is enabled in topic configuration.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

nvartolomei committed:
cloud_storage/tests: test no ts for consumer offsets topic

nvartolomei commented:
I re-opened this at https://github.com/redpanda-data/redpanda/pull/26787 to have a clean slate for review.
